File: dx-api/app/consts/user_level.go
package consts

import (
	"fmt"
	"math"
)

// Level progression consts.
const (
	MaxLevel = 100
	// introExp is the flat EXP cost for Lv.0 → Lv.1, separate from the exponential curve.
	introExp   = 100
	baseExp    = 100
	multiplier = 1.05
)

// UserLevel holds a level number and the cumulative exp required to reach it.
type UserLevel struct {
	Level       int
	ExpRequired int
}

// userLevels is the precomputed level table.
var userLevels []UserLevel

func init() {
	userLevels = generateLevels()
}

// generateLevels builds the full level progression table.
func generateLevels() []UserLevel {
	levels := make([]UserLevel, 0, MaxLevel+1)
	levels = append(levels, UserLevel{Level: 0, ExpRequired: 0})
	levels = append(levels, UserLevel{Level: 1, ExpRequired: introExp})

	cumulative := introExp
	for i := 2; i <= MaxLevel; i++ {
		cumulative += int(math.Floor(baseExp * math.Pow(multiplier, float64(i-2))))
		levels = append(levels, UserLevel{Level: i, ExpRequired: cumulative})
	}

	return levels
}

// GetLevel returns the level for the given cumulative exp.
func GetLevel(exp int) (int, error) {
	if exp < 0 {
		return 0, fmt.Errorf("failed to get level: exp must be non-negative, got %d", exp)
	}
	for i := len(userLevels) - 1; i >= 0; i-- {
		if exp >= userLevels[i].ExpRequired {
			return userLevels[i].Level, nil
		}
	}
	// Unreachable: userLevels[0].ExpRequired is always 0 and exp is non-negative.
	return 0, nil
}

// GetExpForLevel returns the cumulative exp required to reach the given level.
func GetExpForLevel(level int) (int, error) {
	if level < 0 || level > MaxLevel {
		return 0, fmt.Errorf("failed to get exp for level: level must be between 0 and %d, got %d", MaxLevel, level)
	}
	return userLevels[level].ExpRequired, nil
}

File: dx-web/src/consts/user-level.ts
export const MAX_LEVEL = 100;

const BASE_EXP = 100;
const MULTIPLIER = 1.05;

/** Flat EXP cost for Lv.0 → Lv.1, separate from the exponential curve. */
const INTRO_EXP = 100;

export type UserLevel = {
  level: number;
  expRequired: number;
};

function generateLevels(): UserLevel[] {
  const levels: UserLevel[] = [
    { level: 0, expRequired: 0 },
    { level: 1, expRequired: INTRO_EXP },
  ];
  let cumulative = INTRO_EXP;

  for (let i = 2; i <= MAX_LEVEL; i++) {
    cumulative += Math.floor(BASE_EXP * Math.pow(MULTIPLIER, i - 2));
    levels.push({ level: i, expRequired: cumulative });
  }

  return levels;
}

export const USER_LEVELS: UserLevel[] = generateLevels();

export function getLevel(exp: number): number {
  if (exp < 0) {
    throw new Error("exp must be non-negative");
  }
  for (let i = USER_LEVELS.length - 1; i >= 0; i--) {
    if (exp >= USER_LEVELS[i].expRequired) {
      return USER_LEVELS[i].level;
    }
  }
  // Unreachable: USER_LEVELS[0].expRequired is always 0 and exp is non-negative.
  return 0;
}

export function getExpForLevel(level: number): number {
  if (!Number.isInteger(level) || level < 0 || level > MAX_LEVEL) {
    throw new Error(`Level must be an integer between 0 and ${MAX_LEVEL}`);
  }
  return USER_LEVELS[level].expRequired;
}
